### PR TITLE
fixed json parsing for multiple generated JSON code segments, plus co…

### PR DESCRIPTION
**Pull request by rickbraddy-pharma**
# Fixes showstopper issue with parsing LLM-generated code that returns one or multiple code segments in JSON format

Fixes # (issue)

## Who can review?

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:

### CODE REVIEW

**Comment by Kav-K** [comment]
Can someone approve this workflow at least? This is an incredibly important fix. @hwchase17

**Comment by 0ptim** [comment]
I also have many failed runs like #2679 .

My agent sometimes returns:
```
```\n{\n    \"action\": \"Final Answer\",\n    \"action_input\": \"The latest block count is 2950616.\"\n}\n```
```
Instead of:
```
\n{\n    \"action\": \"Final Answer\",\n    \"action_input\": \"The latest block count is 2950616.\"\n}\n
```

**Comment by schinto** [comment]
I guess the issue with the LLM-generated code is not fixed yet.
Using langchain version 0.0.176 the following error occurs when code is returned in triple backticks:

2023-05-22 14:16:40.713 Uncaught app exception
Traceback (most recent call last):
  File "/home/user1/scratch/conda/envs/lcpy39/lib/python3.9/site-packages/langchain/agents/conversational_chat/output_parser.py", line 21, in parse
    cleaned_output, _ = cleaned_output.split("```")
ValueError: too many values to unpack (expected 2)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/user1/scratch/conda/envs/lcpy39/lib/python3.9/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 565, in _run_script
    exec(code, module.__dict__)
  File "/home/user1/langchain-tools/chatbot_streamlit_chat.py", line 116, in <module>
    output, total_tokens, prompt_tokens, completion_tokens = generate_response(
  File "/home/user1/langchain-tools/chatbot_streamlit_chat.py", line 96, in generate_response
    response = chatbot.agent({"input": query})
  File "/home/user1/scratch/conda/envs/lcpy39/lib/python3.9/site-packages/langchain/chains/base.py", line 140, in __call__
    raise e
  File "/home/user1/scratch/conda/envs/lcpy39/lib/python3.9/site-packages/langchain/chains/base.py", line 134, in __call__
    self._call(inputs, run_manager=run_manager)
  File "/home/user1/scratch/conda/envs/lcpy39/lib/python3.9/site-packages/langchain/agents/agent.py", line 947, in _call
    next_step_output = self._take_next_step(
  File "/home/user1/scratch/conda/envs/lcpy39/lib/python3.9/site-packages/langchain/agents/agent.py", line 773, in _take_next_step
    raise e
  File "/home/user1/scratch/conda/envs/lcpy39/lib/python3.9/site-packages/langchain/agents/agent.py", line 762, in _take_next_step
    output = self.agent.plan(
  File "/home/user1/scratch/conda/envs/lcpy39/lib/python3.9/site-packages/langchain/agents/agent.py", line 444, in plan
    return self.output_parser.parse(full_output)
  File "/home/user1/scratch/conda/envs/lcpy39/lib/python3.9/site-packages/langchain/agents/conversational_chat/output_parser.py", line 36, in parse
    raise OutputParserException(f"Could not parse LLM output: {text}") from e
langchain.schema.OutputParserException: Could not parse LLM output: \`\`\`json
{
    "action": "Final Answer",
    "action_input": "\`\`\`python\ndef fibonacci(n):\n    if n <= 0:\n        return []\n    elif n == 1:\n        return [0]\n    elif n == 2:\n        return [0, 1]\n    else:\n        fib_seq = [0, 1]\n        for i in range(2, n):\n            fib_seq.append(fib_seq[-1] + fib_seq[-2])\n        return fib_seq\n\ndef main():\n    fib_numbers = fibonacci(10)\n    print(fib_numbers)\n\nif \_\_name\_\_ == \'\_\_main\_\_\':\n    main()\n\`\`\`\n\nThe algorithm used in this script is an iterative approach to calculate the first 10 Fibonacci numbers. It starts with the base cases of n=1 and n=2, returning [0] and [0, 1] respectively. For n > 2, the script initializes the Fibonacci sequence with the first two numbers (0 and 1) and iterates from the third number to the nth number, appending the sum of the previous two numbers in the sequence. The main() function calls the fibonacci() function with the input 10 and prints the resulting list of Fibonacci numbers."
}
\`\`\`

**Comment by schinto** [comment]
Sorry, I just saw that the fix is not merge in the main branch yet.

**Comment by rickbraddy-pharma** [comment]
Can we please approve the workflow - or is there something else needed to proceed? (this is my first pull request, so happy to adjust as appropriate).

**Review by dev2049 — COMMENTED** [review]
thanks - lets adds some tests for things that this can handle so we don't have any regressions in the future

**Comment by rickbraddy-pharma** [comment]
Resolved Black code formatting issue.

If additional testers are needed, please advise specifics (e.g., which one to modify or a good example to copy/modify)

**Comment by rickbraddy-pharma** [comment]
> thanks - lets adds some tests for things that this can handle so we don't have any regressions in the future

Happy to do so. Can you point me to existing tester you'd suggest that I update?

**Comment by dev2049** [comment]
> > thanks - lets adds some tests for things that this can handle so we don't have any regressions in the future
> 
> Happy to do so. Can you point me to existing tester you'd suggest that I update?

don't think any unit tests exist for this class yet, would make a new file at `tests/unit_tests/agents/conversational_chat/test_output_parser.py` and add there

**Comment by rickbraddy-pharma** [comment]
I have submitted the fix again, this time addressing the "ruff" code format issues.

I have no idea where to start to develop the unit tester...

**Comment by rickbraddy-pharma** [comment]
Figured it out. Will deliver unit tester soon...

**Comment by rickbraddy-pharma** [comment]
Okay. Unit tester is now available.

**Comment by rickbraddy-pharma** [comment]
How should we handle OPEN API keys?

_ ERROR collecting tests/unit_tests/agents/conversational_chat/test_output_parser.py _
tests/unit_tests/agents/conversational_chat/test_output_parser.py:11: in <module>
    llm_chatgpt = ChatOpenAI(temperature=0.3, model_name="gpt-3.5-turbo")
pydantic/main.py:341: in pydantic.main.BaseModel.__init__
    ???
E   pydantic.error_wrappers.ValidationError: 1 validation error for ChatOpenAI
E   __root__
E     Did not find openai_api_key, please add an environment variable `OPENAI_API_KEY` which contains it, or pass  `openai_api_key` as a named parameter. (type=value_error)

File: tests/unit_tests/agents/conversational_chat/test_output_parser.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from __future__ import annotations
+
+from langchain.agents import AgentType, initialize_agent
+from langchain.chat_models import ChatOpenAI
+from langchain.memory import ConversationBufferWindowMemory
+
+# set no tools available (generate only)
```
Comment (eyurtsev) [inline tests/unit_tests/agents/conversational_chat/test_output_parser.py:7]: Encapsulate the test within a function. The test needs to test the agent output parser, so it should not instantiate anything but the parser. 

This test can rely heavily on @pytest.mark.parameterize to test the various parsing scenarios.

Comment (eyurtsev) [reply tests/unit_tests/agents/conversational_chat/test_output_parser.py:7]: I'm on the go right now, but will try to provide a snippet if I manage to do it before I have to turn my laptop off :)

File: tests/unit_tests/agents/conversational_chat/test_output_parser.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+notoolset = []
```
Comment (eyurtsev) [inline tests/unit_tests/agents/conversational_chat/test_output_parser.py:8]: Something along these lines for the file:

```python

from __future__ import annotations

import pytest

from langchain.agents.conversational_chat.output_parser import ConvoOutputParser, OutputParserException


@pytest.mark.parametrize(
    "text, expected",
    [
        ('text', (expected value) ),

    ]
)
def test_parser(text, expected) -> None:
    """Test the ChatOutputParser class."""
    parser = ConvoOutputParser()
    assert parser.parse(text) == expected
    
@pytest.mark.parametrize(
    "text",
    [
        ('texts that lead to parse exception'),

    ]
)
def test_unhappy_path(text) -> None:
    """Test the ChatOutputParser class."""
    parser = ConvoOutputParser()
    with pytest.raises(OutputParserException):
        parser.parse(text)
```

Comment (rickbraddy-pharma) [reply tests/unit_tests/agents/conversational_chat/test_output_parser.py]: Very helpful. Thank you!

The unit tester has been rewritten to account for code and non-code return strings.

**Comment by rickbraddy-pharma** [comment]
The unit tester has been rewritten in accordance with the recommendations.  Please re-run the workflow.

File: tests/unit_tests/agents/conversational_chat/test_output_parser.py (hunk 1 above)
Comment (rickbraddy-pharma) [reply tests/unit_tests/agents/conversational_chat/test_output_parser.py]: I worked with ChatGPT - we rewrote the tester and made sure it passes ruff and mypy checks.

**Comment by rickbraddy-pharma** [comment]
Not sure why the latest commit is a conflict. The entire file was rewritten to replace the prior commit.

**Comment by rickbraddy-pharma** [comment]
Yes, it covers the case where code gets returned within the JSON string, starting with triple backticks and followed by type of code... as shown in the working unit tester.

**Comment by rickbraddy-pharma** [comment]
Who should resolve the conflicts that have cropped up?

**Comment by schinto** [comment]
> Yes, it covers the case where code gets returned within the JSON string, starting with triple backticks and followed by type of code... as shown in the working unit tester.

Sorry, I could not see this nested case in the unit tester.

**Comment by rickbraddy-pharma** [comment]
It was this nested case, where '''python or '''javascript get returned when code is generated that caused the original issues with the output parser. This is what was fixed and why the unit tester was developed to test the code/no-code return cases.

I actually added logging to the output parser, captured actual returned code and non-code parser examples, then included those in the unit tester, to ensure they are correctly structured.

Please let me know if further refinements are needed  to merge.

Thanks
Rick

**Pull request closed without merging**